types/ember__object/proxy.d.ts

```diff
@@ -1,12 +1,33 @@
 import EmberObject from "@ember/object";
+import {
+    UnwrapComputedPropertyGetter,
+    UnwrapComputedPropertyGetters,
+    UnwrapComputedPropertyGettersUndefined
+} from "@ember/object/-private/types";
 
 /**
  * `Ember.ObjectProxy` forwards all properties not defined by the proxy itself
  * to a proxied `content` object.
  */
-export default class ObjectProxy extends EmberObject {
+export default class ObjectProxy<T = object> extends EmberObject {
     /**
      * The object whose properties will be forwarded.
      */
-    content: object;
+    content: T | undefined;
+
+    get<K extends keyof this>(key: K): UnwrapComputedPropertyGetter<this[K]>;
+    get<K extends keyof T>(key: K): UnwrapComputedPropertyGetter<T[K]> | undefined;
+
+    getProperties<K extends keyof this>(
+        list: K[]
+    ): Pick<UnwrapComputedPropertyGetters<this>, K>;
+    getProperties<K extends keyof this>(
+        ...list: K[]
+    ): Pick<UnwrapComputedPropertyGetters<this>, K>;
+    getProperties<K extends keyof T>(
+        list: K[]
+    ): Pick<UnwrapComputedPropertyGettersUndefined<T>, K>;
+    getProperties<K extends keyof T>(
+        ...list: K[]
+    ): Pick<UnwrapComputedPropertyGettersUndefined<T>, K>;
```

Comment:
both `get` and `getProperties` seem to lack support for a property chain

Example:
```ts
const { 'chapters.length': numChapters } = book.getProperties('chapters.length');
```

please make sure to add tests for these cases too

Reply:
Oh - are property chains like this supported with typescript? I wasn't aware that there was away to extract the expected type out of the `chapters.length` string.

I can't see any existing tests for `getProperties` that demonstrate this working with a plain `EmberObject`.

Reply:
> Oh - are property chains like this supported with typescript? I wasn't aware that there was away to extract the expected type out of the `chapters.length` string.

It will not type-check nicely, but it's absolutely part of Ember's public API (which is what these types are intended to describe).

As you've pointed out, we seem to be missing property chain access in other places, although this is where we have support for chains as dependent keys while defining computed properties

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/9a080ecd597b2295b34a4cb6f5609fb08e6632c8/types/ember__object/index.d.ts#L63-L72

Reply:
I assumed this had been left out deliberately, because if you add them you lose the type checking for the properties of the main object.

Currently `.get('unknownProperty')` will cause a TS error. But if we add overloads that accept any string in order to support property chains, we lose that type checking because the overload will always match.

Because of this, I don't tend to use property chaining with `.get` any more. I prefer to sacrifice a few extra lines of code to check any intermediaries for `null/undefined` rather than sacrifice type safety.

The only workaround I can think of is if TS supports a string type that involves a regex match to check if the string contains a `.` character, in which case you could match the overloads to support property chaining without losing the other type checking. But I very much doubt that's possible.

Thoughts?